execution-engine/src/fs.rs

```diff
@@ -0,0 +1,1054 @@
+        };
+        self.inode_table.insert(inode.clone(), node);
+        self.path_table
+            .insert(path.as_ref().to_path_buf(), inode.clone());
+    }
+
+    /// Install a file with content `raw_file_data` and attatch it to `inode`.
+    fn install_file(&mut self, path: impl AsRef<Path>, inode: Inode, raw_file_data: &[u8]) {
+        let file_size = raw_file_data.len();
+        let file_stat = FileStat {
+            device: 0u64.into(),
+            inode: inode.clone(),
+            file_type: FileType::RegularFile,
+            num_links: 0,
+            file_size: file_size as u64,
+            atime: Timestamp::from_nanos(0),
+            mtime: Timestamp::from_nanos(0),
+            ctime: Timestamp::from_nanos(0),
+        };
+        let node = InodeEntry {
+            file_stat,
+            raw_file_data: raw_file_data.to_vec(),
+        };
+        self.inode_table.insert(inode.clone(), node);
+        self.path_table
+            .insert(path.as_ref().to_path_buf(), inode.clone());
+    }
+
+    /// Install a `fd` to the file system. The fd will be of type RegularFile.
+    fn install_fd(
+        &mut self,
+        fd: Fd,
+        inode: Inode,
+        rights_base: &Rights,
+        rights_inheriting: &Rights,
+    ) {
+        let fd_stat = FdStat {
+            file_type: FileType::RegularFile,
+            flags: FdFlags::empty(),
+            rights_base: rights_base.clone(),
+            rights_inheriting: rights_inheriting.clone(),
+        };
+
+        let fd_entry = FileTableEntry {
+            inode: inode.clone(),
+            fd_stat,
+            offset: 0,
+            /// Advice on how regions of the file are to be used.
+            advice: Vec::new(),
+        };
+        self.fd_table.insert(fd.clone(), fd_entry);
+    }
+
+    /// Return a random u32
+    fn random_u32(&self) -> FileSystemError<u32> {
+        let result: [u8; 4] = self
+            .random_get(4)?
+            .as_slice()
+            .try_into()
+            .map_err(|_| ErrNo::Inval)?;
+        Ok(u32::from_le_bytes(result))
+    }
+
+    /// Return a random u64
+    fn random_u64(&self) -> FileSystemError<u64> {
+        let result: [u8; 8] = self
+            .random_get(8)?
+            .as_slice()
+            .try_into()
+            .map_err(|_| ErrNo::Inval)?;
+        Ok(u64::from_le_bytes(result))
+    }
+
+    /// Pick a new fd randomly.
+    fn new_fd(&self) -> FileSystemError<Fd> {
+        loop {
+            let new_fd = self.random_u32()?.into();
+            if !self.fd_table.contains_key(&new_fd) {
+                return Ok(new_fd);
+            }
+        }
+    }
+
+    /// Pick a new inode randomly.
+    fn new_inode(&self) -> FileSystemError<Inode> {
+        loop {
+            let new_inode = Inode(self.random_u64()?);
+            if !self.inode_table.contains_key(&new_inode) {
+                return Ok(new_inode);
+            }
+        }
+    }
+
+    /// Check if `op` is allowed in `fd`
```

Comment:
What's the difference between `op` and `rights`?

Reply:
It should be `rights`.